### PR TITLE
RavenDB-15927 Collection queries are reading their entire input befor…

### PR DESCRIPTION
**Pull request by ayende**
…e they start running

### CODE REVIEW

File: src/Raven.Server/Documents/CollectionRunner.cs [hunk 1]
```diff
@@ -85,36 +85,42 @@ public virtual Task<IOperationResult> ExecuteDelete(string collectionName, long
...
+                                {
+                                    end = true;
+                                    break;
+                                }
 
-                            ids.Enqueue(document.Id);
+                                ids.Enqueue(document.Id);
+
+                                if (ids.Count >= OperationBatchSize)
+                                    break;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/CollectionRunner.cs]: The whole problem is that the collection query is:
1. Not supporting `DocumentField` - should be easy to do?
2. Shouldn't it support paging there, like the others inside 'GetDocuments' ?

File: src/Raven.Server/Documents/CollectionRunner.cs [hunk 2]
```diff
@@ -85,36 +85,42 @@ public virtual Task<IOperationResult> ExecuteDelete(string collectionName, long
                     {
                         foreach (var document in GetDocuments(context, collectionName, startEtag, internalQueryOperationStart, OperationBatchSize, isAllDocs, DocumentFields.Id))
                         {
-                            internalQueryOperationStart++;
+                            using (document)
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/CollectionRunner.cs:88]: We dispose here only the document id and we collect here `OperationBatchSize` which is 1024 IIRC.

File: src/Raven.Server/Documents/CollectionRunner.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/CollectionRunner.cs]: We already request `OperationBatchSize` in the `foreach`

File: src/Raven.Server/Documents/CollectionRunner.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Documents/CollectionRunner.cs:88]: I want to ensure that if we have encrypted database, we return the buffers ASAP

File: src/Raven.Server/Documents/CollectionRunner.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/CollectionRunner.cs]: Yeah, that wasn't handled for `CollectionQueryRunner`, fixing it.

**Comment by ppekrol** [comment]
@ayende I've made some adjustments, can you review?

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 3]
```diff
@@ -697,58 +696,84 @@ public static long ReadLastEtag(Transaction tx)
         }
 
         public IEnumerable<Document> GetDocumentsStartingWith(DocumentsOperationContext context, string idPrefix, string matches, string exclude, string startAfterId,
-            long start, long take, string collection)
+            long start, long take, string collection, DocumentFields fields = DocumentFields.All)
         {
-            foreach (var doc in GetDocumentsStartingWith(context, idPrefix, matches, exclude, startAfterId, start, take))
+            var isAllDocs = collection == Constants.Documents.Collections.AllDocumentsCollection;
+            var requestedDataField = fields.HasFlag(DocumentFields.Data);
+            if (isAllDocs == false && requestedDataField == false)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentsStorage.cs:703]: why do we need this? to read the metadat?

**Review by ayende — COMMENTED** [review]
LGTM

File: src/Raven.Server/Documents/DocumentsStorage.cs (hunk 3 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/DocumentsStorage.cs:703]: Yes

**Comment by arekpalinski** [comment]
LGTM

**Pull request merged**